gcovr/coverage.py

```diff
@@ -314,6 +320,112 @@ def call_coverage(self) -> CoverageStat:
 CovData = Dict[str, FileCoverage]
 
 
+@dataclass
+class DirectoryCoverage:
+    filename: str
+    stats: SummarizedStats
+    children: List[Union[DirectoryCoverage, FileCoverage]]
+    parent_key: str
+
+    @classmethod
+    def new_empty(cls) -> DirectoryCoverage:
+        return cls("", SummarizedStats.new_empty(), list(), "")
+
+    @staticmethod
+    def add_directory_coverage(
+        subdirs: CovData_subdirectories,
+        filename: str,
```

Comment:
No, this is not duplicate info, but i can see why one might interpret that. It is a location to process. Because this is a recursive function, we have to manage the directory location we are putting this statistic within. If the statistic is within A/B/C/D/fileE.cpp, we need to run this function as A/B/C/D/fileE.cpp first, then A/B/C/D, then A/B/C, then A/B, then A, then again at the top level directory. (which is a blank key, or a single Separator-i-forget)

Given the ambiguity, I'll change the name to location and hope that's the best. If not, let me know.

Reply:
Is the `filename` really needed? When the recursion is entered the `newchield` is None. What about using:
```python
    @staticmethod
    def add_directory_coverage(
        subdirs: CovData_subdirectories,
        root_filter: re.Pattern,
        filecov: FileCoverage,
        dircov: DirectoryCoverage = None,
    ) -> None:
        if dircov is None:
            key = DirectoryCoverage.directory_key(filecov.filename, root_filter)
            filecov.parent_key = key
        else:
            key = DirectoryCoverage.directory_key(dircov.dirname, root_filter)
            dircov.parent_key = key
        if key:
            if key not in subdirs:
                subdir = DirectoryCoverage.new_empty()
                subdir.dirname = key
                subdirs[key] = subdir

            if dircov is None:
                subdirs[key].children.append(filecov)
            else:
                if dircov not in subdirs[key].children:
                    subdirs[key].children.append(dircov)

            subdirs[key].stats += SummarizedStats.from_file(filecov)
            DirectoryCoverage.add_directory_coverage(
                subdirs, root_filter, filecov, subdirs[key]
            )
```

The normalizing of the seperator (`force_unix_separator(arg)`) should be done in the `__init__` functions of the `FileCoverage` and the `DirectoryCoverage`.

Reply:
> The normalizing of the seperator (`force_unix_separator(arg)`) should be done in the `__init__` functions of the `FileCoverage` and the `DirectoryCoverage`.

@macetw What do you think about this? Than the writer can rely on unix separators and it would be much clearer.